src/garage/torch/algos/vpg.py

```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
             discount=self.discount)
         return np.mean(undiscounted_returns)
 
-    def _train(self, obs, actions, rewards, advantages):
-        r"""Train the algorithm with minibatch.
+    def _train(self, obs_flat, actions_flat, rewards_flat, returns_flat,
```

Comment:
it's strange for this function to take rewards, returns, and advantages, because returns can always be calculated from rewards, and advantages can always be calculated using returns and the value_function

i was confused because we usually reserve `process_samples` for operations which happen in the CPU, and use `train` for operations which occur in the computation graph (e.g. the GPU if you are running in GPU mode). this class blurs the line, and it's difficult to tell where the np.ndarrays stop and the torch.Tensors start.

Reply:
I'll take out returns and baselines from the `process_samples` and put these in `train_once`.
Because it has to compute `returns` and `advantages` using the whole trajectories, these values should be computed before separating values into mini-batches.